minidump-processor/src/processor.rs

```diff
@@ -738,3 +768,76 @@ where
 
     Ok(state)
 }
+
+/// A cursor that allows a memory list to be used like a byte stream
```

Comment:
My commentary here is purely advisory in my status as maintainer emeritus, but are you sure this abstraction pulls its weight here? In my experience a minidump either has a single memory region that contains the memory containing the instruction pointer or it doesn't. I can't recall ever seeing an instance where there was an adjacent memory region that would have been useful to continue reading in this manner.

Does this code set you up for future success in some way with other planned work?

For comparison's sake, here's the Rust port of my `get-minidump-instructions` tool that I used many times for doing this analysis in an ad-hoc manner:
https://github.com/rust-minidump/rust-minidump/blob/4f6eae86dd16f0aaf0b6a8056989d76b691a9934/minidump-tools/src/lib.rs#L216-L247

It simply tries to find a memory region containing the instruction pointer, and uses that if available. I cannot recall ever running into a situation where having this abstraction would have produced more useful results.

Reply:
Some of this code is potentially reusable by third-party integrations like minidump-debugger. I could imagine this sort of logic being useful for debugging Full Memory Minidumps, but I'm not sure what the stream looks like in practice for that situation (i.e. can a contiguous region be fragmented across entries?).

Reply:
> I could imagine this sort of logic being useful for debugging Full Memory Minidumps, but I'm not sure what the stream looks like in practice for that situation

We have code for reading the memory streams from full memory dumps, each memory region is a contiguous chunk of bytes:
https://github.com/rust-minidump/rust-minidump/blob/a6a486b8992008c6865ee70fc68421c38e45ef42/minidump/src/minidump.rs#L1896-L1915

Reply:
> > I did this mostly because I needed some trait anyway to make it easy to unit test without synthesizing my own fake memory list. I originally had a GetMemoryAtAddress single-function trait and changed it to this. I guess I will just go back to that
> 
> I'm always sympathetic to anything that makes it easier to write tests! I think the [`SynthMinidump`](https://docs.rs/minidump-synth/latest/minidump_synth/) bits we have are full-featured enough to make this straightforward. You can see an example of creating a synthetic x86-64 dump in this test:
> 
> https://github.com/rust-minidump/rust-minidump/blob/38edd05c27588f573a447d89d07bdba252f601d9/minidump/src/minidump.rs#L6561-L6574
> 
> The processor tests have a function to build a minimal minidump that they add additional info to:
> 
> https://github.com/rust-minidump/rust-minidump/blob/a6a486b8992008c6865ee70fc68421c38e45ef42/minidump-processor/tests/test_processor.rs#L136-L149
> 
> Calling `SynthMinidump::add_memory` automatically adds the memory region you provide to the minidump and also to the memory list stream.
> 
> There are tests with examples of adding exception info:
> 
> https://github.com/rust-minidump/rust-minidump/blob/38edd05c27588f573a447d89d07bdba252f601d9/minidump/src/minidump.rs#L6697-L6709
> 
> All of this is built atop the [`test-assembler` crate](https://docs.rs/test-assembler/latest/test_assembler/), which gives you a bunch of ways to string together arbitrary data for testing purposes. The stackwalker tests are a good point of reference, they set up the stack contents to test specific unwind scenarios, you could do virtually the same thing to create memory regions with specific instructions for your tests:
> 
> https://github.com/rust-minidump/rust-minidump/blob/a6a486b8992008c6865ee70fc68421c38e45ef42/minidump-processor/src/stackwalker/amd64_unittest.rs#L116-L130

Thank you for all the info here. I will probably use this for the unit testing for the stuff I build on top of this feature (like non-canonical address handling and other things in the list shared by @gabrielesvelto)